File: leetcode/1-100/44-wildcard-matching/bench/wildcard_matching.c
// LeetCode #44 bench mirror — C, the greedy two-pointer matcher (★).
//
// Mirrors bench/wildcard_matching.kara: one cursor in each of s and p, with star/matched
// scalars for O(1) backtracking. Build s ("abc" repeated) and a multi-star pattern p ONCE,
// then punch one s slot per iteration and fold the boolean into a rolling checksum. Same
// workload + sink as every other mirror.

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

static int is_match(const unsigned char *s, int64_t n, const unsigned char *p, int64_t m) {
    int64_t i = 0, j = 0, star = -1, matched = 0;
    while (i < n) {
        if (j < m && (p[j] == '?' || p[j] == s[i])) {
            i++;
            j++;
        } else if (j < m && p[j] == '*') {
            star = j;
            matched = i;
            j++;
        } else if (star != -1) {
            matched++;
            i = matched;
            j = star + 1;
        } else {
            return 0;
        }
    }
    while (j < m && p[j] == '*') {
        j++;
    }
    return j == m;
}

int main(void) {
    const int64_t total = 300000;
    const int64_t modulus = 1000000007;
    const int64_t n = 240;

    unsigned char *s = (unsigned char *)malloc((size_t)n);
    for (int64_t a = 0; a < n; a++) {
        s[a] = (unsigned char)('a' + (a % 3));
    }
    const int64_t m = 33; // 8 * "*abc" + "*"
    unsigned char *p = (unsigned char *)malloc((size_t)m);
    int64_t pi = 0;
    for (int64_t g = 0; g < 8; g++) {
        p[pi++] = '*';
        p[pi++] = 'a';
        p[pi++] = 'b';
        p[pi++] = 'c';
    }
    p[pi++] = '*';

    int64_t acc = 0;
    for (int64_t k = 0; k < total; k++) {
        s[k % n] = (unsigned char)('a' + (k % 4));
        int64_t bit = is_match(s, n, p, m) ? 1 : 0;
        acc = (acc * 131 + bit) % modulus;
    }

    free(s);
    free(p);
    printf("%lld\n", (long long)acc);
    return 0;
}

File: leetcode/1-100/44-wildcard-matching/bench/wildcard_matching.rs
// LeetCode #44 bench mirror — Rust, the greedy two-pointer matcher (★).
//
// Mirrors bench/wildcard_matching.kara exactly: one cursor in each of s and p, with
// star/matched scalars for O(1) backtracking. Build s and a multi-star pattern p once, punch
// one s slot per iteration, fold the boolean into a rolling checksum. Same workload + sink as
// every other mirror.

fn is_match(s: &[u8], p: &[u8]) -> bool {
    let n = s.len() as i64;
    let m = p.len() as i64;
    let mut i = 0i64;
    let mut j = 0i64;
    let mut star = -1i64;
    let mut matched = 0i64;
    while i < n {
        if j < m && (p[j as usize] == b'?' || p[j as usize] == s[i as usize]) {
            i += 1;
            j += 1;
        } else if j < m && p[j as usize] == b'*' {
            star = j;
            matched = i;
            j += 1;
        } else if star != -1 {
            matched += 1;
            i = matched;
            j = star + 1;
        } else {
            return false;
        }
    }
    while j < m && p[j as usize] == b'*' {
        j += 1;
    }
    j == m
}

fn main() {
    let total: i64 = 300000;
    let modulus: i64 = 1000000007;
    let n: i64 = 240;

    let mut s: Vec<u8> = (0..n).map(|a| b'a' + (a % 3) as u8).collect();
    let mut p: Vec<u8> = Vec::new();
    for _ in 0..8 {
        p.push(b'*');
        p.push(b'a');
        p.push(b'b');
        p.push(b'c');
    }
    p.push(b'*');

    let mut acc: i64 = 0;
    let mut k: i64 = 0;
    while k < total {
        s[(k % n) as usize] = b'a' + (k % 4) as u8;
        let bit = if is_match(&s, &p) { 1 } else { 0 };
        acc = (acc * 131 + bit) % modulus;
        k += 1;
    }

    println!("{}", acc);
}
